ipi/utils/parsing.py

```diff
@@ -0,0 +1,222 @@
+                                data_processed = "\n".join(data_frame)
+                            data_list.append(data_processed)
+                            data_frame = []
+
+                        # Found a new step, update current_step and initialize the list for data
+                        step_list.append(int(matches[0][1]))
+                        if property_name == "":
+                            property_name = matches[0][0]
+                        elif property_name != matches[0][0]:
+                            raise ValueError(
+                                f"Inconsistent property {matches[0][0]} found in extras containing {property_name}"
+                            )
+                    else:
+                        data_frame.append(line)
+
+                if len(data_frame) > 0:
+                    try:
+                        data_processed = np.loadtxt(data_frame)
+                    except:
+                        is_array = False
+                        data_processed = "\n".join(data_frame)
+                    data_list.append(data_processed)
+                if is_array:
+                    data_list = np.array(data_list).squeeze()
+                return {
+                    "step": np.asarray(step_list, dtype=int),
+                    property_name: data_list,
+                }
+            else:
+                ret = read_file(
+                    format,
+                    file_handle,
+                    dimension=dimension,
+                    units=units,
+                    cell_units=cell_units,
+                )
+
+                frame = ase.Atoms(
+                    ret["atoms"].names,
+                    positions=ret["atoms"].q.reshape((-1, 3)) * bohr2angstrom,
+                    cell=ret["cell"].h.T * bohr2angstrom,
+                    pbc=True,
+                )
+
+                # parse comment to get the property
+                matches = comment_regex.findall(ret["comment"])
+
+                # get what we have found
+                if len(matches) >= 2:
+                    what = matches[-2][0]
+                else:  # defaults to reading positions
+                    what = "positions"
+
+                # ... and the step
+                matches = step_regex.findall(ret["comment"])
+                if len(matches) >= 1:
+                    frame.info["step"] = int(matches[-1][0])
+
+                # if we have forces, set positions to zero (that data is missing!) and set forces instead
+                if what == "forces":
+                    # set forces and convert to eV/angstrom
+                    frame.positions *= 0
+                    frame.arrays["forces"] = ret["atoms"].q.reshape(
+                        (-1, 3)
+                    ) * unit_to_user("force", "ev/ang", 1.0)
+
+                frames.append(frame)
+
+        except EOFError:
+            break
+        except:
+            raise
+
+    return frames
```

Comment:
Great feature. I and I am sure many others have a collection of many scripts doing pieces of this. Really glad it is now in a single